<a href="https://colab.research.google.com/github/PixelGif786/CAP4770-rvd-1231/blob/main/pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Pandas

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/pandas-logo-2.png)

<font size='0.85em'>(logo by [Marc Garcia](https://github.com/pandas-dev/pandas/blob/master/web/pandas/static/img/pandas.svg))</font> 


> A suggestion: This worksheet is organized in a way that, hopefully, best presents the topics and not in a way that is the most useful for reference. As you progress through this worksheet, your future you will find it useful if you create your own notes so that you can quickly find the relevant information. These notes might take the form of a separate Google Colab notebook or handwritten notes. In addition, just the activity of taking notes will help you retain the information.

The Pandas Library is built on top of Numpy and is designed to make working with data fast and easy. Like Numpy, the library includes data structures and functions to manipulate that data.

As we learned in the Numpy Notebook, we need to load in the library before we can use it.

In [1]:
from pandas import Series, DataFrame
import numpy as np
import pandas as pd

Let's dissect the code above.

The `Series` and `DataFrame` datatypes are commonly used so we import them directly with

```
from pandas import Series, DataFrame
```

For all other datatypes and functions in the library, the `pd` prefix is commonly used so we import that with

```
import pandas as pd
```

## Series
A series is a 1d array-like object

Let's consider the heights (in cm) of the members of the Japan's Women's Basketball Team at the 2020 Olympics (they won the silver medal).      

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/basketball2020.png)
We can create a series in a number of ways.

For example, we can create it directly from a Python list

In [2]:
japan = [183, 185, 167, 162, 165, 174, 173, 181, 167, 183, 185, 182] # a Python list
athletesHeight = Series(japan) # converted to a Pandas Series

We could also have done

In [3]:
athletesHeight = Series([183, 185, 167, 162, 165, 174, 173, 181, 167, 183, 185, 182])

In either case we can see the value of the Series `athletesHeight`

In [4]:
 athletesHeight

0     183
1     185
2     167
3     162
4     165
5     174
6     173
7     181
8     167
9     183
10    185
11    182
dtype: int64

You are probably familiar with arrays in programming languages and this works in a similar way. The left number is the index and the right the value. And we can find the value at a particular index by the usual:

In [5]:
athletesHeight[3]

162

#### specifying indices
Instead of the index 0, 1, 2, 3 ... you can specify your own index values. For example, we can label them 'Moeko Nagaoka', 'Maki Takada', 'Naho Miyoshi' ... etc. 

In [6]:
athletes2 = Series(japan, index = ['Moeko Nagaoka', 'Maki Takada', 'Naho Miyoshi', 
                                   'Rui Machida', 'Nako Motohashi', 'Nanaka Todo', 
                                   'Saki Hayashi','Evelyn Mawuli', 'Saori Miyazaki', 'Yuki Miyazawa', 'Himawari Akaho', 
                                   'Monica Okoye'])
athletes2

Moeko Nagaoka     183
Maki Takada       185
Naho Miyoshi      167
Rui Machida       162
Nako Motohashi    165
Nanaka Todo       174
Saki Hayashi      173
Evelyn Mawuli     181
Saori Miyazaki    167
Yuki Miyazawa     183
Himawari Akaho    185
Monica Okoye      182
dtype: int64

The names we see are not another column of the data. We can see the shape of athletes2 by:

In [7]:
athletes2.shape

(12,)

This shows that athletes2 is a one dimensional matrix and that dimension has a length of 12. So the names we see are not values in a column but rather the indices.


Let's use the index to get the Height of Himawari Akaho:

In [8]:
athletes2['Himawari Akaho']

185


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)
## DataFrame
DataFrames are **the most important data structure of Pandas** and are simply
a table or spreadsheet like structure.  A DataFrame represents a table like:
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ioniq5s.png)

Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | 256 | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | 113 | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

### Creating a DataFrame by Hand

A common way to create a DataFrame is to use a Python dictionary as follows:

In [9]:
 cars = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [256, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, 113, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}


and now we can create a DataFrame from the `cars` Python dictionary:

In [10]:
df = DataFrame(cars)
df


Make Drive  Range  Battery_kwH  PeakPower_kW  MPGe  0-60
0         Ioniq 5   AWD    256         77.4           239    98   5.0
1   Tesla Model 3   AWD    315         80.0           298   113   3.1
2      Rivian R1T   AWD    316        135.0           562    70   3.0
3         VW ID.4   RWD    260         82.0           260   107   7.7
4  F150 Lightning   AWD    230        110.0           318    68   4.0

Prior to my life with Pandas, I would represent a table like the above one as:

In [11]:
prePandas = [{'Make': 'Ioniq 5', 'Drive': 'AWD', 'Range': 256, 'Battery_kwH': 77.4, 'PeakPower_kW': 98, 'MPGe': 98, '0-60': 5.0},
             {'Make': 'Tesla Model 3', 'Drive': 'AWD', 'Range': 315, 'Battery_kwH': 80.0, 'PeakPower_kW': 113,'MPGe': 113, '0-60': 3.1},
             {'Make': 'Rivian R1T', 'Drive': 'AWD', 'Range': 316, 'Battery_kwH': 135.0, 'PeakPower_kW': 562,'MPGe': 70, '0-60': 3.0},
             {'Make': 'VW ID.4', 'Drive': 'RWD', 'Range': 260, 'Battery_kwH': 82.0, 'PeakPower_kW': 260,'MPGe': 107, '0-60': 7.7},
             {'Make': 'F150 Lightning', 'Drive': 'AWD', 'Range': 230, 'Battery_kwH': 110.0, 'PeakPower_kW': 318,'MPGe': 68, '0-60': 4.0}]


prePandas[0]['Make']

'Ioniq 5'

In the prePandas scheme the data is organized first by rows. That seemed logical to me since each row represents an object and is how we organize data in an SQL database. In the Pandas representation the data is organized by columns. 



![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Reading data from a csv file:

A CSV file (comma separated values) is a common data format for data science files. As| the name suggests, the data values in a row are separated by commas. For example, the data file to represent the table:


Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | 256 | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | 113 | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0


would be

```
Make,Drive,Range,Battery_kwH,PeakPower_kW,MPGe,0-60
Ioniq 5,AWD,256,77.4,239,98,5.0
Tesla Model 3,AWD,315,80,298,113,3.1
Rivian R1T,AWD,316,135,562,70,3.0 
VW ID.4,RWD,260,82,260,107,7.7
F150 Lightning,AWD,230,110,318,68,4.0

```


As the name suggests, we use the `pd.read_csv` function to read a csv file.  `pd.read_csv` can read a csv file from either your local machine or the web. Let's start with the web.

### Reading a CSV file from the web.
To read a file from the web, we simply provide a URL:


In [12]:
evs = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv')
evs

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]

Sometimes the csv file has a header row as was the case in the example above. That file starts with the line
```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
```
as we can see by using `curl`



In [13]:
!curl -s https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv |  head

Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
Tesla,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
BMW,iX3,6.8,180,360,206,560,Yes,RWD,Type 2 CCS,SUV,D,5,68040
Honda,e,9.5,145,170,168,190,Yes,RWD,Type 2 CCS,Hatchback,B,4,32997
Lucid,Air,2.8,250,610,180,620,Yes,AWD,Type 2 CCS,Sedan,F,5,105000
Volkswagen,e-Golf,9.6,150,190,168,220,Yes,FWD,Type 2 CCS,Hatchback,C,5,31900
Peugeot,e-208,8.1,150,275,164,420,Yes,FWD,Type 2 CCS,Hatchback,B,5,29682
Tesla,Model 3 Standard Range Plus,5.6,225,310,153,650,Yes,RWD,Type 2 CCS,Sedan,D,5,46380


**An aside**

We can preface any Unix command with a bang (!) to have it execute in our Notebook.  This is amazingly handy:



In [14]:
!ls

sample_data


In [ ]:
!pwd

/content


### Files with No Header Row
Sometimes the csv file does not have a header row. So for example, data might start on the very first line of the file

```
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen ,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar ,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
```

In that case you specify the names of the columns using the `names` parameter:

In [15]:
columnNames = ['Brand','Model','AccelSec','TopSpeed_KmH','Range_Km','Efficiency_WhKm',
               'FastCharge_KmH','RapidCharge','PowerTrain','PlugType','BodyStyle',
               'Segment','Seats','PriceEuro']

evs2 = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean_NoHeader.csv', names=columnNames)
evs2

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla     Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                         ID.3 Pure      10.0           160   
2      Polestar                                 2       4.7           210   
3           BMW                              iX3        6.8           180   
4         Honda                                e        9.5           145   
..           ...                              ...       ...           ...   
98       Nissan                       Ariya 63kWh       7.5           160   
99         Audi     e-tron S Sportback 55 quattro       4.5           210   
100      Nissan               Ariya e-4ORCE 63kWh       5.9           200   
101      Nissan   Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                 M-Byte 95 kWh 2WD       7.5           190   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
98        330              191            440         Yes        FWD   
99        335              258            540         Yes        AWD   
100       325              194            440         Yes        AWD   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  
0    Type 2 CCS      Sedan       D      5      55480  
1    Type 2 CCS  Hatchback       C      5      30000  
2    Type 2 CCS   Liftback       D      5      56440  
3    Type 2 CCS        SUV       D      5      68040  
4    Type 2 CCS  Hatchback       B      4      32997  
..          ...        ...     ...    ...        ...  
98   Type 2 CCS  Hatchback       C      5      45000  
99   Type 2 CCS        SUV       E      5      96050  
100  Type 2 CCS  Hatchback       C      5      50000  
101  Type 2 CCS  Hatchback       C      5      65000  
102  Type 2 CCS        SUV       E      5      62000  

[103 rows x 14 columns]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

### Reading a CSV file the local machine
First, let's get that file onto our local machine:

In [19]:
curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv > evs.csv

SyntaxError: ignored

Hmm. That didn't work. Can you fix the error and rerun that cell?

---
Now we can specify the local file using `pd=read_csv`

In [ ]:
ev3 = pd.read_csv('evs.csv')
ev3

FileNotFoundError: ignored

Suppose we want that file in a data directory. Let's go ahead and create the directory and move the file there.

In [ ]:
!mkdir data
!mv evs.csv data

mv: cannot stat 'evs.csv': No such file or directory


Now when we load the file we need to give more of a path:

In [ ]:
evs4 = pd.read_csv('data/evs.csv')
evs4

FileNotFoundError: ignored


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Missing Data


In real machine learning tasks, we often encounter missing values.

### Missing Data in Files
For example, suppose we didn't know The Polestar 2's acceleration, the BMW iX3's top speed and the Honda's model name. In that case our CSV file would start


```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,,210,400
BMW ,iX3 ,6.8,,360
Honda ,,9.5,145,
```

with the double comma on the Polestar, BMW and Honda lines representing the missing data. When we read that file.



In [ ]:
evs5 = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small.csv')
evs5

Brand                          Model  AccelSec  TopSpeed_KmH  \
0        Tesla   Model 3 Long Range Dual Motor       4.6         233.0   
1   Volkswagen                       ID.3 Pure      10.0         160.0   
2     Polestar                               2       NaN         210.0   
3          BMW                            iX3        6.8           NaN   
4        Honda                             NaN       9.5         145.0   
5        Lucid                            Air        2.8         250.0   
6   Volkswagen                         e-Golf        9.6         150.0   
7      Peugeot                          e-208        8.1         150.0   
8        Tesla     Model 3 Standard Range Plus       5.6         225.0   
9         Audi                      Q4 e-tron        6.3         180.0   
10    Mercedes                  EQC 400 4MATIC       5.1         180.0   
11      Nissan                           Leaf        7.9         144.0   
12     Hyundai            Kona Electric 64 kWh       7.9         167.0   
13         BMW                             i4        4.0         200.0   
14     Hyundai                  IONIQ Electric       9.7         165.0   
15  Volkswagen                      ID.3 Pro S       7.9         160.0   
16     Porsche                  Taycan Turbo S       2.8         260.0   
17  Volkswagen                          e-Up!       11.9         130.0   
18          MG                           ZS EV       8.2         140.0   
19        Mini                      Cooper SE        7.3         150.0   

    Range_Km  
0      450.0  
1      270.0  
2      400.0  
3      360.0  
4        NaN  
5      610.0  
6      190.0  
7      275.0  
8      310.0  
9      400.0  
10     370.0  
11     220.0  
12     400.0  
13     450.0  
14     250.0  
15     440.0  
16     375.0  
17     195.0  
18     220.0  
19     185.0

We see that missing values are now the floating point values NaN meaning Not a Number. This NaN value is used even in columns that do not contain floating point values. For example, in row 4 above, Honda has NaN in the model name column. Wes McKinney, the developer of Pandas calls NaN a *sentinel* value that is easily detected and indicates a missing value.

### Special Characters Representing NaN in the Data File

Sometimes, special characters are used in a data file to represent missing values. For example, sometimes a dash is used ...

Brand|Model|AccelSec|TopSpeed_KmH|Range_Km
:--- | :--- | ---: | ---: | ---: |
Tesla |Model 3 Long Range Dual Motor|4.6|233|450
Volkswagen |ID.3 Pure|10.0|160|270
Polestar |2|-|210|400
BMW |iX3 |6.8|-|360
Honda |-|9.5|145|

The associated data file would look like

```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,-,210,400
BMW ,iX3 ,6.8,-,360
Honda ,-,9.5,145,
```

If we do not convert these to `NaN`, these dashes will create havoc with future calculations:

``` 
TypeError: can't multiply sequence by non-int of type 'float'
```

When we read in the csv file we need to convert the dashes to `NaN` by using the `na_values` parameter in `read_csv`. As you can see in the following example, our data file has the dashes 

In [ ]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small-dash.csv

and we can convert those to `NaN` when we read the file

In [ ]:
evdash = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small-dash.csv',
                     na_values='-')
evdash[:6]

### Specifying Missing Values by hand
Suppose we didn't know the range of the Hyundai Ioniq 5 and the MPGe of the Tesla.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ioniq5s.png)

Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | - | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | - | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

In that case we can create a dataframe like:

In [ ]:
cars3 = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [np.nan, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, np.nan, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}
        
carz = DataFrame(cars3)
carz

where `np.nan` is Numpy's NaN.  We can also use Python's `None`:


In [ ]:
cars3 = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [None, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, None, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}
        
carz = DataFrame(cars3)
carz

In addition to reading CSV files, there are many other ways of reading in data including from SQL databases, mongoDB, and webpages. See the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/) for details.



![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Indices
In all the examples above, when we created a DataFrame, an auto-generated, auto-incrementing index was generated. So our initial data might look like:


Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | - | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | - | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

But the DataFrame looks like ...

In [ ]:
carz
   

with the 0, 1, 2 ... indices added (the column on the left). This makes it handy when we want to access a particular row.

In [ ]:
carz.loc[4]

As you can see in the above code, we access a particular row (or rows) with the `loc` method.

Sometimes the rows in our data already have a unique identifier. For example in a data file for U.S. states each state might have the unique 2 character state code


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/grayStates.png)

State|Name|Pop2022|Pop2021|Pop2010|TotalArea|LandArea
:--- | :--- | ---: | ---: | ---: | ---: | ---: | 
AK| Alaska| 720763|724357|713910|665384|570640.95
AL| Alabama|4949697|4934193|4785437|52420|50645.33
AR| Arkansas|3042017|3033946|2921964|53179|52035.48
AZ| Arizona|7640796|7520103|6407172|113990|113594.08

In that case we can use the unique identifier, in this case the 2 letter state code, as our index. 

In [ ]:
states = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/usStates.csv',
                     index_col='State')
states

We can access a particular row, say the info for New Mexico by 

In [ ]:
states.loc['NM']

You can access a range of rows by the standard Python method:

In [ ]:
states[:5]

or


In [ ]:
states[11:16]

or using these index values:

In [ ]:
states['IA': 'KS']


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Retrieving particular columns
We can get a column of a DataFrame by using the column name:

In [ ]:
states[['Pop2022']]

and we can get multiple columns by passing a list of column names

In [ ]:
states[['Pop2022', 'Pop2010']]

### Returning rows with loc

As we have seen we can retrieve rows using the loc function. 

We can also get rows that match a specific criterion. For example, all states whose 2022 population exceeded 10 million. 

In [ ]:
populousStates = states.loc[states['Pop2022'] > 10000000] 
populousStates

Or let's say we are interested in the states that lost population between 2021 and 2022.

In [ ]:
states.loc[states['Pop2022'] < states['Pop2021']]

States that are over 10 million population and have lost population in 2022:

In [ ]:
states.loc[(states['Pop2022'] < states['Pop2021']) & (states['Pop2022'] > 10000000)]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Creating new columns

#### State Density

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/stateDensity.png)

Let's say we want to add a new column, Density, that gives the number of people per square mile of land area. So the Formula is



$$Density=\frac{Pop2022}{LandArea}$$

We can do that with


In [ ]:
states['Density'] = states['Pop2022'] / states['LandArea']
states[:5][['Name', 'Density']]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Sorting

Density is nice but it would be nice if we ordered the DataFrame by Density...

In [ ]:
states.sort_values(by=['Density'])

That is better, but perhaps we would like to order by densest states first and only show the top five ...
 

In [ ]:
states.sort_values(by=['Density'], ascending=False)[:5]

Finally, let's add a column, growth, that shows the percent change in population from 2021 to 2022.  That formula would be

$$Growth=(\frac{Pop2022}{Pop2021} - 1) \times 100  $$

In [ ]:
states['Growth'] = (states['Pop2022'] / states['Pop2021'] - 1) * 100
states[:10]

And let's find the 5 fastest growing states:

In [ ]:
states.sort_values(by=['Growth'], ascending=False)[:5]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Descriptive Statistics on DataFrames
One handy function is `describe`

In [ ]:
states.describe()

Alternatively, we could retrieve a specific statistic:


In [ ]:
states['Pop2022'].mean()

or find out how many states lost population in 2022 using `count`:

In [ ]:
states.loc[states['Pop2022'] < states['Pop2021']][['Pop2022']].count()

Or get the total population of the United States (minus the District of Columbia)

In [ ]:
states['Pop2022'].sum()

or to make that more readable

In [ ]:
totalPop = "{:,}".format(states['Pop2022'].sum())
print('Total Population: ' + totalPop)

What percentage of the U. S. population is accounted for by the top 5 most populous states?

In [ ]:
statesByPopulation = states.sort_values(by=['Pop2022'], ascending=False)[:10][['Pop2022']]
percent = (statesByPopulation.sum() / states['Pop2022'].sum()) * 100
print("The top 10 most populous states account for {:4.2f} percent of the U.S. Population" .format(percent['Pop2022']))

## Summary Statistics and Axes

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ax.png)
Suppose I have customers of my vinyl record shop rate different artists

|Customer | Taylor Swift | Miranda Lambert | Carrie Underwood | Nicki Minaj | Ariana Grande |
|:-----------|:------:|:------:|:---------:|:------:|:--------:|
|Jake|5|-|5|2|2|
|Clara|2|-|-|4|5|
|Kelsey|5|5|5|2|-|
|Angelica|2|3|-|5|5|
|Jordyn|2|1|-|5|-|

First we will read in the file ...

In [ ]:
ratings = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/ratings.csv', index_col=0)
ratings

We can get the mean rating of each artist by:


In [ ]:
ratings.mean()

Note that the summary statistics ignore NaN entries. The mean rating for Miranda Lambert is computed just on the people that rated her.

Many descriptive statistics functions take an optional parameter `axis` that tells which axis to reduce over. If we want the mean ratings for each **customer** instead of each artist we can do:

In [ ]:
ratings.mean(axis=1)



Sweet! So `axis=1` means reduce by rows and `axis=0` means reduce by columns:

In [ ]:
ratings.mean(axis=0)

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## List of Descriptive Statistics
(from the book *Python for Data Analysis*)

Method | Description
:-- | :--
`count` | Number of non-NaN values
`describe` | A set of common summary statistics
`min, max` | compute minimum and maximum values
`argmin, argmax` | compute index locations of minimum and maximum values
`sum` | Sum the values
`mean` | Mean of values
`median` | Median of values
`std` | Sample standard deviation

So, for example, the lowest rating for each artist:


In [ ]:
ratings.min()


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

# <font color='#EE4C2C'>You Try ...</font> 
Ok, it is time for you to try out what you just learned. Let us start with the electric vehicle datafile we have already seen:

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv
```
When we loaded the datafile before, it worked fine for those examples, but you may encounter an error here, that will require you to modify the `read_csv`.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/evBanner.png)

In [ ]:
! curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv
cars = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv',
                     na_values='-')
cars

Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
Tesla,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
BMW,iX3,6.8,180,360,206,560,Yes,RWD,Type 2 CCS,SUV,D,5,68040
Honda,e,9.5,145,170,168,190,Yes,RWD,Type 2 CCS,Hatchback,B,4,32997
Lucid,Air,2.8,250,610,180,620,Yes,AWD,Type 2 CCS,Sedan,F,5,105000
Volkswagen,e-Golf,9.6,150,190,168,220,Yes,FWD,Type 2 CCS,Hatchback,C,5,31900
Peugeot,e-208,8.1,150,275,164,420,Yes,FWD,Type 2 CCS,Hatchback,B,5,29682
Tesla,Model 3 Standard Range Plus,5.6,225,310,153,650,Yes,RWD,Type 2 CCS,Sedan,D,5,46380
Audi,Q4 e-tron,6.3,180,400,193,540,Yes,AWD,Type 2 CCS,SUV,D,5,55000
Mercedes,EQC 400 4MATIC,5.1,180,370,216,440,Yes,AWD,Type 2 CCS,SUV,D,5,69484
Nissan,Leaf,7.9,144,220,164,230,Yes,FWD,Typ

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0         450              161           940.0         Yes        AWD   
1         270              167           250.0         Yes        RWD   
2         400              181           620.0         Yes        AWD   
3         360              206           560.0         Yes        RWD   
4         170              168           190.0         Yes        RWD   
..        ...              ...             ...         ...        ...   
101       375              232           450.0         Yes        AWD   
102       400              238           480.0         Yes        AWD   
103       505              299           337.0         Yes        AWD   
104       529              446           250.0         Yes        AWD   
105       514              299           309.0         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  
0    Type 2 CCS      Sedan       D      5    55480.0  
1    Type 2 CCS  Hatchback       C      5    30000.0  
2    Type 2 CCS   Liftback       D      5    56440.0  
3    Type 2 CCS        SUV       D      5    68040.0  
4    Type 2 CCS  Hatchback       B      4    32997.0  
..          ...        ...     ...    ...        ...  
101  Type 2 CCS  Hatchback       C      5    65000.0  
102  Type 2 CCS        SUV       E      5    62000.0  
103  Type 2 CCS     Pickup       N      6    75000.0  
104  Type 2 CCS      Truck       E      5        NaN  
105  Type 2 CCS      Truck       E      5        NaN  

[106 rows x 14 columns]

Some of the columns are based on kilometers:

* TopSpeed_KmH	
* Range_Km	
* Efficiency_WhKm	
* FastCharge_KmH

and we are first going to add USA mile-centric columns.

## <font color='#EE4C2C'>1. TopSpeed_MPH</font>

Add a new column `TopSpeed_MPH` computed from the `TopSpeed_KmH` column, and just to check, display the first 10 rows. (The Lucid Air should have a top speed slightly over 155 MPH)


In [ ]:
cars['TopSpeed_MPH'] = cars['TopSpeed_KmH'] * 0.6214
#cars[:5][['Model', 'TopSpeed_MPH']]
cars

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0         450              161           940.0         Yes        AWD   
1         270              167           250.0         Yes        RWD   
2         400              181           620.0         Yes        AWD   
3         360              206           560.0         Yes        RWD   
4         170              168           190.0         Yes        RWD   
..        ...              ...             ...         ...        ...   
101       375              232           450.0         Yes        AWD   
102       400              238           480.0         Yes        AWD   
103       505              299           337.0         Yes        AWD   
104       529              446           250.0         Yes        AWD   
105       514              299           309.0         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH  
0    Type 2 CCS      Sedan       D      5    55480.0      144.7862  
1    Type 2 CCS  Hatchback       C      5    30000.0       99.4240  
2    Type 2 CCS   Liftback       D      5    56440.0      130.4940  
3    Type 2 CCS        SUV       D      5    68040.0      111.8520  
4    Type 2 CCS  Hatchback       B      4    32997.0       90.1030  
..          ...        ...     ...    ...        ...           ...  
101  Type 2 CCS  Hatchback       C      5    65000.0      124.2800  
102  Type 2 CCS        SUV       E      5    62000.0      118.0660  
103  Type 2 CCS     Pickup       N      6    75000.0      114.9590  
104  Type 2 CCS      Truck       E      5        NaN      106.2594  
105  Type 2 CCS      Truck       E      5        NaN      111.8520  

[106 rows x 15 columns]


## <font color='#EE4C2C'>2. Range and FastCharge_MPH</font>

Add a new column `Range` computed from the `Range_Km` column that shows the range in miles.

Also add a column `FastCharge_MPH computed from `FastCharge_KmH`. This will indicate how many miles can you go on one hour of charging. 

Just to check, display the first 10 rows. (The Tesla Model 3 Long Range Dual Motor should have a range of nearly 280 miles and charge at a rate of 584 miles per hour of fast charging.)

In [ ]:
cars['Range'] = cars['Range_Km'] * 0.6214

cars['FastCharge_MPH'] = (cars['FastCharge_KmH'] * 0.6214)

cars[:10]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla  Model 3 Long Range Dual Motor       4.6           233   
1  Volkswagen                      ID.3 Pure      10.0           160   
2    Polestar                              2       4.7           210   
3         BMW                            iX3       6.8           180   
4       Honda                              e       9.5           145   
5       Lucid                            Air       2.8           250   
6  Volkswagen                         e-Golf       9.6           150   
7     Peugeot                          e-208       8.1           150   
8       Tesla    Model 3 Standard Range Plus       5.6           225   
9        Audi                      Q4 e-tron       6.3           180   

   Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0       450              161           940.0         Yes        AWD   
1       270              167           250.0         Yes        RWD   
2       400              181           620.0         Yes        AWD   
3       360              206           560.0         Yes        RWD   
4       170              168           190.0         Yes        RWD   
5       610              180           620.0         Yes        AWD   
6       190              168           220.0         Yes        FWD   
7       275              164           420.0         Yes        FWD   
8       310              153           650.0         Yes        RWD   
9       400              193           540.0         Yes        AWD   

     PlugType  BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH    Range  \
0  Type 2 CCS      Sedan       D      5    55480.0      144.7862  279.630   
1  Type 2 CCS  Hatchback       C      5    30000.0       99.4240  167.778   
2  Type 2 CCS   Liftback       D      5    56440.0      130.4940  248.560   
3  Type 2 CCS        SUV       D      5    68040.0      111.8520  223.704   
4  Type 2 CCS  Hatchback       B      4    32997.0       90.1030  105.638   
5  Type 2 CCS      Sedan       F      5   105000.0      155.3500  379.054   
6  Type 2 CCS  Hatchback       C      5    31900.0       93.2100  118.066   
7  Type 2 CCS  Hatchback       B      5    29682.0       93.2100  170.885   
8  Type 2 CCS      Sedan       D      5    46380.0      139.8150  192.634   
9  Type 2 CCS        SUV       D      5    55000.0      111.8520  248.560   

   FastCharge_MPH  
0         584.116  
1         155.350  
2         385.268  
3         347.984  
4         118.066  
5         385.268  
6         136.708  
7         260.988  
8         403.910  
9         335.556

## <font color='#EE4C2C'>3. Efficiency_MkwH</font>

The DataFrame has a column `Efficiency_WhKm` which indicates how many watt hours does it take to go one kilometer. One common measure of efficieny is how many miles can you go on one kilowatt hour (kwH). So please add this column to the DataFrame. (The Tesla Model 3 Standard Range Plus should have an efficiency of around 4 miles per kilowatt hour.) 


In [ ]:
cars['Efficiency_MkwH'] = (cars['Efficiency_WhKm'] / 38.25)

In [ ]:
cars

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0         450              161           940.0         Yes        AWD   
1         270              167           250.0         Yes        RWD   
2         400              181           620.0         Yes        AWD   
3         360              206           560.0         Yes        RWD   
4         170              168           190.0         Yes        RWD   
..        ...              ...             ...         ...        ...   
101       375              232           450.0         Yes        AWD   
102       400              238           480.0         Yes        AWD   
103       505              299           337.0         Yes        AWD   
104       529              446           250.0         Yes        AWD   
105       514              299           309.0         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH     Range  \
0    Type 2 CCS      Sedan       D      5    55480.0      144.7862  279.6300   
1    Type 2 CCS  Hatchback       C      5    30000.0       99.4240  167.7780   
2    Type 2 CCS   Liftback       D      5    56440.0      130.4940  248.5600   
3    Type 2 CCS        SUV       D      5    68040.0      111.8520  223.7040   
4    Type 2 CCS  Hatchback       B      4    32997.0       90.1030  105.6380   
..          ...        ...     ...    ...        ...           ...       ...   
101  Type 2 CCS  Hatchback       C      5    65000.0      124.2800  233.0250   
102  Type 2 CCS        SUV       E      5    62000.0      118.0660  248.5600   
103  Type 2 CCS     Pickup       N      6    75000.0      114.9590  313.8070   
104  Type 2 CCS      Truck       E      5        NaN      106.2594  328.7206   
105  Type 2 CCS      Truck       E      5        NaN      111.8520  319.3996   

     FastCharge_MPH  Efficiency_MkwH  
0          584.1160         4.209150  
1          155.3500         4.366013  
2          385.2680         4.732026  
3          347.9840         5.385621  
4          118.0660         4.392157  
..              ...              ...  
101        279.6300         6.065359  
102        298.2720         6.222222  
103        209.4118         7.816993  
104        155.3500        11.660131  
105        192.0126         7.816993  

[106 rows x 18 columns]

## <font color='#EE4C2C'>4. New DataFrame</font>

Create a new DataFrame, `bevs` (battery electric vehicles as opposed to PHEVs) that contain only the following columns from the current DataFrame

* Brand	
* Model	
* PowerTrain	
* AccelSec
* TopSpeed_MPH	
* Range	
* Efficiency_MkwH

```
'Brand', 'Model', 'PowerTrain', 'AccelSec', 'TopSpeed_MPH', 'Range', 'Efficiency_MkwH'
```


In [ ]:
bevs = DataFrame(cars)

In [ ]:
bevs[['Brand', 'Model', 'PowerTrain', 'AccelSec', 'TopSpeed_MPH', 'Range', 'Efficiency_MkwH']]

Brand                            Model PowerTrain  AccelSec  \
0         Tesla    Model 3 Long Range Dual Motor        AWD       4.6   
1    Volkswagen                        ID.3 Pure        RWD      10.0   
2      Polestar                                2        AWD       4.7   
3           BMW                              iX3        RWD       6.8   
4         Honda                                e        RWD       9.5   
..          ...                              ...        ...       ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance        AWD       5.1   
102       Byton                M-Byte 95 kWh 2WD        AWD       7.5   
103      Rivian                              R1T        AWD       3.0   
104         GMC                        Hummer EV        AWD       3.0   
105        Ford                   F150 Lightning        AWD       3.0   

     TopSpeed_MPH     Range  Efficiency_MkwH  
0        144.7862  279.6300         4.209150  
1         99.4240  167.7780         4.366013  
2        130.4940  248.5600         4.732026  
3        111.8520  223.7040         5.385621  
4         90.1030  105.6380         4.392157  
..            ...       ...              ...  
101      124.2800  233.0250         6.065359  
102      118.0660  248.5600         6.222222  
103      114.9590  313.8070         7.816993  
104      106.2594  328.7206        11.660131  
105      111.8520  319.3996         7.816993  

[106 rows x 7 columns]

## <font color='#EE4C2C'>5. Most and least efficient</font>
Using the bevs DataFrame, what are the 5 most efficient evs ordered by the most efficient first?


In [ ]:
bevs.sort_values(by=['Efficiency_MkwH'], ascending=False)[:5]

Brand                Model  AccelSec  TopSpeed_KmH  Range_Km  \
104       GMC            Hummer EV       3.0           171       529   
105      Ford       F150 Lightning       3.0           180       514   
103    Rivian                  R1T       3.0           185       505   
84   Mercedes         EQV 300 Long      10.0           140       330   
90       Audi  e-tron S 55 quattro       4.5           210       320   

     Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain    PlugType  \
104              446           250.0         Yes        AWD  Type 2 CCS   
105              299           309.0         Yes        AWD  Type 2 CCS   
103              299           337.0         Yes        AWD  Type 2 CCS   
84               273           290.0         Yes        FWD  Type 2 CCS   
90               270           510.0         Yes        AWD  Type 2 CCS   

    BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH     Range  \
104     Truck       E      5        NaN      106.2594  328.7206   
105     Truck       E      5        NaN      111.8520  319.3996   
103    Pickup       N      6    75000.0      114.9590  313.8070   
84        SPV       N      7    70631.0       86.9960  205.0620   
90        SUV       E      5    93800.0      130.4940  198.8480   

     FastCharge_MPH  Efficiency_MkwH  
104        155.3500        11.660131  
105        192.0126         7.816993  
103        209.4118         7.816993  
84         180.2060         7.137255  
90         316.9140         7.058824

What are the five most inefficient evs ordered by least efficient first?

In [ ]:
bevs.sort_values(by=['Efficiency_MkwH'], ascending=True)[:5]

Brand                        Model  AccelSec  TopSpeed_KmH  Range_Km  \
48  Lightyear                          One      10.0           150       575   
8       Tesla  Model 3 Standard Range Plus       5.6           225       310   
14    Hyundai               IONIQ Electric       9.7           165       250   
41    Hyundai         Kona Electric 39 kWh       9.9           155       255   
74       Sono                         Sion       9.0           140       225   

    Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain    PlugType  \
48              104           540.0         Yes        AWD  Type 2 CCS   
8               153           650.0         Yes        RWD  Type 2 CCS   
14              153           210.0         Yes        FWD  Type 2 CCS   
41              154           210.0         Yes        FWD  Type 2 CCS   
74              156           270.0         Yes        FWD  Type 2 CCS   

    BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH    Range  \
48   Liftback       F      5   149000.0        93.210  357.305   
8       Sedan       D      5    46380.0       139.815  192.634   
14   Liftback       C      5    34459.0       102.531  155.350   
41        SUV       B      5    33971.0        96.317  158.457   
74  Hatchback       C      5    25500.0        86.996  139.815   

    FastCharge_MPH  Efficiency_MkwH  
48         335.556         2.718954  
8          403.910         4.000000  
14         130.494         4.000000  
41         130.494         4.026144  
74         167.778         4.078431

## <font color='#EE4C2C'>6. fastest
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/porsche.png)

The Porsche 911 997 Turbo S is among the world's fastest production combustion cars with a 0-60 acceleration of 2.9 seconds. (In our dataset `AccelSec` measures the same thing.) Which electric cars have better acceleration than this? 

In [ ]:
bevs2 = bevs.loc[bevs['AccelSec'] < 3.0]
bevs2

Brand                Model  AccelSec  TopSpeed_KmH  Range_Km  \
5     Lucid                  Air       2.8           250       610   
16  Porsche       Taycan Turbo S       2.8           260       375   
51    Tesla             Roadster       2.1           410       970   
59    Tesla  Model S Performance       2.5           261       505   
81    Tesla  Model X Performance       2.8           250       440   

    Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain    PlugType  \
5               180           620.0         Yes        AWD  Type 2 CCS   
16              223           780.0         Yes        AWD  Type 2 CCS   
51              206           920.0         Yes        AWD  Type 2 CCS   
59              188           550.0         Yes        AWD      Type 2   
81              216           480.0         Yes        AWD      Type 2   

   BodyStyle Segment  Seats  PriceEuro  TopSpeed_MPH    Range  FastCharge_MPH  \
5      Sedan       F      5   105000.0      155.3500  379.054         385.268   
16     Sedan       F      4   180781.0      161.5640  233.025         484.692   
51    Cabrio       S      4   215000.0      254.7740  602.758         571.688   
59  Liftback       F      5    96990.0      162.1854  313.807         341.770   
81       SUV       F      7   102990.0      155.3500  273.416         298.272   

    Efficiency_MkwH  
5          4.705882  
16         5.830065  
51         5.385621  
59         4.915033  
81         5.647059

## <font color='#EE4C2C'>7. Refrigerators
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/frigTesla.png)

Representative Thomas Massie (R-KY) at a hearing questioning Secretarty Pete Buttigieg says the energy (kWh) required to home charge an electric vehicle for a year is 25 times greater than running the average refrigerator. He also said "It would take four times as much electricity to charge the average household's cars as the average household uses on air conditioning." And he said, "that means the average household would use twice as much electricity charging one of their cars as they would use for all of the air conditioning that they use for the entire year." Is this true? 

Since air conditioner use is quite varied in different regions of the country, perhaps a better measure would be the energy used for air conditioning and heating combined.

You are going to need to do some web searching to find the typical energy use of household these household items.

What do you think about the usefulness of this comparing household appliances to EVs when making government policy?  


Government policies are required to be made in favor of the people. Since most people are not able to fully understand the scientific measures and units when it come for measuring the power consumption like kWh and most of them would not get what is that unit. For a better understanding of the common people the government should promote the comparisons based on their daily life uses like the use of refrigerators and use of air conditioners .
Precisely these units of measurements are not accurate and cannot be fully used as a formal unit of measurements as these are only comparable values and are not fixed. The average household consumption of electricity is not a definite value but people can easily grasp what it means charging of an EV as compared to use of AC or fridge. Their uses are different across different regions and in the year. For a colder region the use of fridge and AC is minimal but for a hotter region the use is grater , thus more consumption of power in that region. So, government should aim to make the common comparison units taking into account the region and other factors.



## <font color='#EE4C2C'>8. Global Adoption of EVs

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/earth.png)

The data file we are using for this is

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv
```

This data is from the [International Energy Agency](https://www.iea.org/data-and-statistics/data-product/global-ev-outlook-2022).

The first step in data mining work is to examine the dataset, which often is not in the best format to directly answer our questions. That is the case with this file. You have the Pandas skills, but you will need to spend a bit of time looking at the datafile and understanding it, so you can write some code to meet the need.

So what is our need? We are interested in what percent of new car sales are evs in 2021. We are interested in historical data, not predictions. We would like to see the top five countries with the best adoption percent and the bottom five. Have appropriate labels indicating the list of top 5 and bottom 5. 



In [ ]:
from pandas import Series, DataFrame
import pandas as pd
data = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv',
                     na_values='-')

In [ ]:
cars1=DataFrame(data)
#cars1=DataFrame(data,index=['powertrain'])
cars1

cars1['powertrain']
cars1[['powertrain','year']]
( (cars1['powertrain']=='EV') & (cars1['year']== 2021) ).sum()

total = 915/7016
total

0.13041619156214368

## <font color='#EE4C2C'>9. Adoption of EVs Worldwide Part 2
In the U.S. in 2021 Electric Vehicles account for 2.5% of total new car sales. Is this higher or lower than the average (mean) 2021 figures in our dataset?

Also, how many countries have a higher adoption rate than the U.S?

In addition to cells showing your work, please have some readable content.

In [ ]:
value1 = 7016*0.025
value1
(cars1['year']== 2021).mean()
#915>175.4


0.13041619156214368

In [ ]:
cars1[['region','value']]
cars1[cars1['region']== 'USA'].mean()
cars1[cars1['value'] > 3054615.549] 

<ipython-input-42-3f035fc17bcb>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  cars1[cars1['region']== 'USA'].mean()


region          category parameter  mode powertrain  year   unit  \
933   China    Projection-APS  EV stock  Cars        BEV  2020  stock   
990   China  Projection-STEPS  EV stock  Cars        BEV  2020  stock   
1036  China        Historical  EV stock  Cars        BEV  2020  stock   
1041  China        Historical  EV stock  Cars        BEV  2021  stock   
1136  China  Projection-STEPS  EV stock  Cars        BEV  2021  stock   
...     ...               ...       ...   ...        ...   ...    ...   
6990  World  Projection-STEPS  EV sales  Cars        BEV  2030  sales   
6991  World  Projection-STEPS  EV sales  Cars       PHEV  2030  sales   
6996  World  Projection-STEPS  EV stock  Vans        BEV  2030  stock   
6998  World  Projection-STEPS  EV stock  Cars        BEV  2030  stock   
6999  World  Projection-STEPS  EV stock  Cars       PHEV  2030  stock   

            value  
933     3512477.0  
990     3512477.0  
1036    3512477.0  
1041    6246490.0  
1136    6246490.0  
...           ...  
6990   21977040.0  
6991    5719586.0  
6996   16099721.0  
6998  129716784.0  
6999   45116820.0  

[101 rows x 8 columns]

The electric vehicles accountting for 2.5% of total new cars sales are 175.4 total, this percentage is not higher than the avg mean of 2021 0.025<0.13